wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+	 *
+	 * Each query has its own return value.
+	 *
+	 * @var mixed
+	 */
+	private $return_value;
+
+	/**
+	 * Variable to keep track of nested transactions level.
+	 *
+	 * @var int
+	 */
+	private $transaction_level = 0;
+
+	/**
+	 * Value returned by the last exec().
+	 *
+	 * @var mixed
+	 */
+	private $last_exec_returned;
+
+	/**
+	 * The PDO fetch mode passed to query().
+	 *
+	 * @var mixed
+	 */
+	private $pdo_fetch_mode;
+
+	/**
+	 * @var WP_SQLite_Information_Schema_Builder
+	 */
+	private $information_schema_builder;
+
+	/**
+	 * Constructor.
+	 *
+	 * Create PDO object, set user defined functions and initialize other settings.
+	 * Don't use parent::__construct() because this class does not only returns
+	 * PDO instance but many others jobs.
+	 *
+	 * @param string $db_name The database name. In WordPress, the value of DB_NAME.
+	 * @param PDO|null $pdo The PDO object.
+	 */
+	public function __construct( string $db_name, ?PDO $pdo = null ) {
+		$this->db_name = $db_name;
+		if ( ! $pdo ) {
+			if ( ! is_file( FQDB ) ) {
+				$this->prepare_directory();
+			}
+
+			$locked      = false;
+			$status      = 0;
+			$err_message = '';
+			do {
+				try {
+					$options = array(
+						PDO::ATTR_ERRMODE           => PDO::ERRMODE_EXCEPTION,
+						PDO::ATTR_STRINGIFY_FETCHES => true,
+						PDO::ATTR_TIMEOUT           => 5,
+					);
+
+					$dsn = 'sqlite:' . FQDB;
+					$pdo = new PDO( $dsn, null, null, $options );
+				} catch ( PDOException $ex ) {
+					$status = $ex->getCode();
+					if ( self::SQLITE_BUSY === $status || self::SQLITE_LOCKED === $status ) {
+						$locked = true;
+					} else {
+						$err_message = $ex->getMessage();
+					}
+				}
+			} while ( $locked );
+
+			if ( $status > 0 ) {
+				$message                = sprintf(
+					'<p>%s</p><p>%s</p><p>%s</p>',
+					'Database initialization error!',
+					"Code: $status",
+					"Error Message: $err_message"
+				);
+				$this->is_error         = true;
+				$this->error_messages[] = $message;
+				return;
+			}
+		}
+
+		WP_SQLite_PDO_User_Defined_Functions::register_for( $pdo );
+
+		// MySQL data comes across stringified by default.
+		$pdo->setAttribute( PDO::ATTR_STRINGIFY_FETCHES, true );
+
+		$this->pdo = $pdo;
+
+		$this->information_schema_builder = new WP_SQLite_Information_Schema_Builder(
+			$this->db_name,
+			array( $this, 'execute_sqlite_query' )
+		);
+		$this->information_schema_builder->ensure_information_schema_tables();
+
+		// Load MySQL grammar.
+		if ( null === self::$grammar ) {
+			self::$grammar = new WP_Parser_Grammar( require self::GRAMMAR_PATH );
+		}
+
+		// Fixes a warning in the site-health screen.
+		// @TODO: It is not clear how this fixes a warning or if it serves a
+		//        different purpose. We should also make this use PDO and
+		//        consider providing that information lazily via a getter.
+		$this->client_info = SQLite3::version()['versionString'];
+
+		$this->pdo->query( 'PRAGMA foreign_keys = ON' );
```

Comment:
Is this consistent with how WordPress MySQL tables work by default? I'm thinking about SET modes and MyISAM vs innodb

Reply:
The default today is InnoDB and in that case, when you add FKs, they just work. In SQLite, however, they need to be turned on. WordPress doesn't use FKs, but some plugins may do so.

Reply:
Let's keep an eye on this – we might need to make it configurable by the caller. WordPress tables may be altered to use MyISAM and plugins may create custom tables with different storage engines.

Reply:
I added a note about this to a new issue: https://github.com/Automattic/sqlite-database-integration/issues/5